File: kmeans_ci.m


function [CI,K, centroids]=kmeans_ci(X,stand,prop,nclus,nsim);

% [CI,K]=kmeans_ci(X,stand,prop,nclus,nsim);
%
%  
%
% Input
% 'X': input matrix to be classified (rows are classified)
% 'stand': option for standardising the matrix (by columns) = 'm' for just
% removing the long-term mean and ='s' for standardising to zero mean and
% unit variance. if stand is empty, the matrix is not standardized.
% 'prop': scalar giving the proportion of variance to be retained in the
% EOF pre-filtering step. if prop is empty, no EOF-prefiltering is done
% 'nclus': number of cluster
% 'nsim' : number of different partition simulated (typically nsim= 50-100)
%
% Outputs
% 'CI': scalar giving the classifiability index
% 'K': row vector giving the cluster for each row of input matrix 'X'.
%
% ref. Michelangeli et al., JAS, 1995 (1230-1246)
%
 
if ~isempty(stand);
    X=stan(X,stand);
end

if ~isempty(prop);
    [U,S,V]=svd(X,0);
    s=diag(S).^2;
    sc=s/sum(s);
    a=find(cumsum(sc)> prop);
    a=a(1);
    X=U(:,1:a)*S(1:a,1:a);
end

[r,c]=size(X);
disp('Time steps and EOF accounting for the varfrac:')
size(X)
centroids_all = cell(nsim, 1);

for i=1:nsim;
    k(:,i)=kmeans(X,nclus,'Maxiter',1000,'EmptyAction','drop');
    for j=1:nclus;
       mean_cluster(j,:)=mean(X(find(k(:,i)==j),:));
    end
    %Guardar los centroides en la matriz 3D
    centroids_all{i} = mean_cluster;
    
    mean_cluster2=stan(mean_cluster','s'); clear mean_cluster
    MC(:,i)=mean_cluster2(:); % centroids stored in MC matrix
end
centroids = centroids_all{end};
% Anomaly correlation between partition P and Q
for i=1:nclus;
    for j=1:nsim;
        sample1=MC(((i-1)*c)+1:i*c,j);
        a=find(j~=[1:nsim]);
        sample2=reshape(MC(:,a),c,(nsim-1)*nclus); clear ACC
        ACC=(1/(c-1))*sample1'*sample2;
        ACC=reshape(ACC,nclus,nsim-1);
        ACCmax(i,j)=min(max(ACC)); % mean similarity for cluster i of the simulation j
    end
end

part=find(mean(ACCmax)==max(mean(ACCmax)));
CI=mean(mean(ACCmax)); % Classification index for the partition on 'nclus' clusters
if length(part) > 1; part=part(1); end
K=k(:,part); % best partition that maximize the ACC with the nsim-1 other partitions

File: kmeans_ci.py
from sklearn.decomposition import TruncatedSVD
import numpy as np
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import numpy as np
import cv2

def stand(y, opt):
    my = np.nanmean(y, axis=0)
    
    sty = np.nanstd(y, axis=0,ddof=1)
   # Evitar divisiones por desviaciones estándar muy pequeñas
    sty[sty < 1e-8] = 1.0  # Umbral mínimo para la desviación estándar
    my = np.tile(my,  (y.shape[0], 1))
    sty = np.tile(sty, (y.shape[0], 1))
    
    # opt='m'
    if opt == 'm':
            x = y - my
    elif opt =='s':
        x = (y - my) / sty
    return x

#vERIFIACAR FUNCION STAND

def kmeans_ci_py (X, stan, prop, nclus, nsim):
    if stan is not None:
        X = stand(X, stan)
    
    
    U, S, Vt = np.linalg.svd(X,full_matrices=False)
    s = (S)**2
    sc = s/np.sum(s)
    a = np.where(np.cumsum(sc) > prop)
    b = a[0][0]
    X = U[:, :b+1] @np.diag(S[:b+1])
        
    r, c = X.shape
    # print('Time steps and EOF ccouting for the varfrac: ')
    # print(X.shape)
    
    k = np.zeros((r,nsim), dtype=int)
    MC = np.zeros((nclus * c, nsim))
    # nsim=10
    for i in range(nsim):
        # print(i)
        kmeans = KMeans(n_clusters=nclus,init='k-means++', max_iter=10, n_init=10)
        # algoritmo = KMeans(n_clusters = a,  init = 'k-means++',max_iter = 300, n_init = 100, random_state=0)
        kmeans.fit(X)

        k[:,i] = kmeans.predict(X)
        mean_cluster = kmeans.cluster_centers_
        # print(k[:,i])
        # mean_cluster = np.zeros((nclus,c))
        
        # for j in range(nclus):
        #     mean_cluster[j,:] = np.mean(X[np.flatnonzero(k[:,i] == j), :], axis=0)
        #     # ass=X[np.flatnonzero(k[:,i] == j), :]
           
        mean_cluster2 = stand(mean_cluster.T, 's') 
        
        aux=cv2.flip(mean_cluster2.T,0)
        MC[:, i] = aux.flatten()
            
    ACCmax = np.zeros((nclus, nsim))
            
    for i in range(nclus):
        for j in range(nsim):
            sample1 = MC[(i * c):(i + 1)* c, j]
            a=np.where(np.arange(nsim) !=j)[0]
            sample2 = MC[:, a].reshape(c, (nsim - 1) * nclus,order='F')
            ACC = (1 / (c - 1)) * np.dot(sample1.T, sample2)   
            ACC = ACC.reshape(nclus, nsim - 1,order='F')
            ACCmax[i, j] = np.min(np.max(ACC, axis=0))
    
    mean_ACCmax = np.mean(ACCmax, axis=0)
    # Selección de la mejor partición
    part = np.where(mean_ACCmax == np.max(mean_ACCmax))[0]
    CI = np.mean(ACCmax)
    if len(part) >= 1:
            part = part[0]
    return CI, k[:, part]